apps/passport-server/migrations/58_password_reset_limit.sql

```diff
@@ -0,0 +1,98 @@
+CREATE TABLE rate_limit_types(
+  event_type VARCHAR NOT NULL,
+  hourly_limit INT NOT NULL,
+  PRIMARY KEY(event_type)
+);
+
+CREATE TABLE rate_limit_buckets(
+  event_type VARCHAR NOT NULL,
+  event_id VARCHAR NOT NULL,
+  remaining INT NOT NULL,
+  last_refill TIMESTAMPTZ,
+  PRIMARY KEY(event_type, event_id),
+  FOREIGN KEY(event_type) REFERENCES rate_limit_types
+);
+
+-- Allow 10 attempts per hour to verify tokens for the same email
+INSERT INTO rate_limit_types VALUES('CHECK_EMAIL_TOKEN', 10);
+-- Allow 10 attempts per hour to request a new email token, either
+-- as part of creating or re-creating an account.
+INSERT INTO rate_limit_types VALUES('REQUEST_EMAIL_TOKEN', 10);
+
+-- MIT License
+
+-- Copyright (c) 2017 Fabian Flechtmann
+
+-- Permission is hereby granted, free of charge, to any person obtaining a copy
+-- of this software and associated documentation files (the "Software"), to deal
+-- in the Software without restriction, including without limitation the rights
+-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+-- copies of the Software, and to permit persons to whom the Software is
+-- furnished to do so, subject to the following conditions:
+
+-- The above copyright notice and this permission notice shall be included in all
+-- copies or substantial portions of the Software.
+
+-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+-- SOFTWARE.
+
+-- This implements a "token bucket" approach to rate limiting, where each
+-- request causes a counter to decrease, where that counter is also re-filled
+-- based on time elapsed since the last refill. This allows us to ensure that
+-- requests are limited to a certain rate, without hard cut-offs around expiry
+-- or without needing to track each individual request.
+--
+-- Based on code here: https://github.com/fafl/token-bucket-postgres
+
+DROP FUNCTION IF EXISTS take_token(VARCHAR, VARCHAR);
+CREATE OR REPLACE FUNCTION take_token (event_type VARCHAR(100), event_id VARCHAR(100)) RETURNS boolean AS $$
```

Comment:
suggestion: the `VARCHAR(100)` type probably makes sense to propagate into your table definitions above, otherwise you create surface area for unexpected behavior. eg. it's not out of the question for someone to expect a bucket id to be able to be a concatenation of several hashes or guids. I'm also not sure that 100 makes sense as a limit, or if there even should be a limit at all here.